src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,247 @@
+    if (!_connections[hostNameString]) {
+        char portString[6];
+        snprintf(portString, sizeof(portString), "%d", ntohs(port));
+        nw_endpoint_t endpoint = nw_endpoint_create_host(hostName, portString);
+        if (!endpoint) {
+            MTR_LOG_ERROR("%@ failed to create endpoint for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_parameters_t params = nw_parameters_create_secure_udp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
+        if (!params) {
+            MTR_LOG_ERROR("%@ failed to create udp parameters", self);
+            return;
+        }
+        nw_connection_t connection = nw_connection_create(endpoint, params);
+        if (!connection) {
+            MTR_LOG_ERROR("%@ failed to create connection for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_connection_set_queue(connection, sSharedResolverQueue);
+        nw_connection_set_path_changed_handler(connection, ^(nw_path_t _Nonnull path) {
+            nw_path_status_t status = nw_path_get_status(path);
+            if (status == nw_path_status_satisfied) {
+                MTR_LOG_INFO("%@ path is satisfied", self);
+                std::lock_guard lock(sConnectivityMonitorLock);
+                [self _callHandler];
+            }
+        });
+        nw_connection_set_viability_changed_handler(connection, ^(bool viable) {
+            if (viable) {
+                std::lock_guard lock(sConnectivityMonitorLock);
+                MTR_LOG_INFO("%@ connectivity now viable", self);
+                [self _callHandler];
+            }
+        });
+        nw_connection_start(connection);
+    }
+}
+
+static void _resolveReplyCallback(
+    DNSServiceRef sdRef,
+    DNSServiceFlags flags,
+    uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode,
+    const char * fullname,
+    const char * hosttarget,
+    uint16_t port, /* In network byte order */
+    uint16_t txtLen,
+    const unsigned char * txtRecord,
+    void * context)
+{
+    auto * connectivityMonitor = (__bridge MTRDeviceConnectivityMonitor *) context;
+    [connectivityMonitor handleResolvedHostname:hosttarget port:port error:errorCode];
+}
+
+- (void)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
+{
+    std::lock_guard lock(sConnectivityMonitorLock);
+
+    MTRDeviceConnectivityMonitorHandler handlerCopy = [handler copy];
+    _monitorHandler = handlerCopy;
+    _handlerQueue = queue;
+
+    // If there's already a resolver running, just return
+    if (_resolver) {
+        MTR_LOG_INFO("%@ connectivity monitor updated handler", self);
+        return;
+    }
+
+    MTR_LOG_INFO("%@ start connectivity monitoring for %@ (%lu monitoring objects)", self, _instanceName, static_cast<unsigned long>(sConnectivityMonitorCount));
+
+    _resolver = [MTRDeviceConnectivityMonitor _sharedResolverConnection];
```

Comment:
Do we (or why do we) need to do our own DNS-SD resolve here by hand, instead of using / hooking into the operational resolve within the CHIP stack?

Reply:
MTROperationalBrowser also deals with some of the same things... overall it feels like we could use some cohesion between these things.

Reply:
Hm that's a lot of moving code around. I think we can certainly look into doing that instead, once we have this running and tests verify this strategy works well to trigger resubscriptions.